## Gaussian Elimination

### 1. Pivoting vs. No Pivoting

a. Using gaussian elimination, rounding to three decimal places each step..

In [68]:
def linearsolver(A,b):
    n = len(A)
    M = A

    i = 0
    for x in M:
        x.append(b[i])
        i += 1

    for k in range(n):
        for i in range(k,n):
            if abs(M[i][k]) > abs(M[k][k]):
                M[k], M[i] = M[i],M[k]
            else:
                pass

    for j in range(k+1,n):
        q = float(M[j][k]) / M[k][k]
        for m in range(k, n+1):
            M[j][m] -=  q * M[k][m]

    x = [0 for i in range(n)]

    x[n-1] =float(M[n-1][n])/M[n-1][n-1]
    for i in range (n-1,-1,-1):
        z = 0
        for j in range(i+1,n):
            z = z  + float(M[i][j])*x[j]
    x[i] = float(M[i][n] - z)/M[i][i]
    print(x)




In [89]:
import numpy as np

j=[0.143,0.357,2.01,-5.173]
k=[-1.31,.911,1.99,-5.458]
l=[11.2,-4.3,-.605,4.415]
print('initial matrix-  \n ' + str(np.asmatrix([j,k,l])))
j=[round((1/.143)*i,3) for i in j]
k=[round((1/-1.31)*i,3) for i in k]
l=[round((1/11.2)*i,3) for i in l]
print("now we have 1's in left most column -  \n " + str(np.asmatrix([j,k,l])))
k=np.subtract(k,j)
l=np.subtract(l,j)
print("R2-R1 , R3-R1   \n " + str(np.asmatrix([j,k,l])))
k=[round((1/-3.192)*i,3) for i in k]
l=[round((1/-2.881)*i,3) for i in l]

print("continuing with 2nd column   \n " + str(np.asmatrix([j,k,l])))
l=np.subtract(l,k)
l=[round((1/.019)*i,3) for i in l]
print( "... \n" + str(np.asmatrix([j,k,l])) + " \n now we have all 1's on diagonal so we will start back substitution \n ")

l_alpha=[round((-4.879)*i,3) for i in l]
k=np.add(k,l_alpha)
l_beta=[round((-14.056)*i,3) for i in l]
j=np.add(j,l_beta)
print("R2 + -.4879*R3   \n  R1 + -14.056*R3 " + str(np.asmatrix([j,k,l])))
k_alpha=[round((-2.497)*i,3) for i in k]
j=np.add(j,k_alpha)
print("R1 + -2.497*R2   \n " + str(np.asmatrix([j,k,l])))

initial matrix-  
 [[  0.143   0.357   2.01   -5.173]
 [ -1.31    0.911   1.99   -5.458]
 [ 11.2    -4.3    -0.605   4.415]]
now we have 1's in left most column -  
 [[  1.      2.497  14.056 -36.175]
 [  1.     -0.695  -1.519   4.166]
 [  1.     -0.384  -0.054   0.394]]
R2-R1 , R3-R1   
 [[  1.      2.497  14.056 -36.175]
 [  0.     -3.192 -15.575  40.341]
 [  0.     -2.881 -14.11   36.569]]
continuing with 2nd column   
 [[  1.      2.497  14.056 -36.175]
 [ -0.      1.      4.879 -12.638]
 [ -0.      1.      4.898 -12.693]]
... 
[[  1.      2.497  14.056 -36.175]
 [ -0.      1.      4.879 -12.638]
 [  0.      0.      1.     -2.895]] 
 now we have all 1's on diagonal so we will start back substitution 
 
R2 + -.4879*R3   
  R1 + -14.056*R3 [[ 1.     2.497  0.     4.517]
 [-0.     1.     0.     1.487]
 [ 0.     0.     1.    -2.895]]
R1 + -2.497*R2   
 [[ 1.     0.     0.     0.804]
 [-0.     1.     0.     1.487]
 [ 0.     0.     1.    -2.895]]


This gives solution [0.804, 1.487, -2.895] which is different from the actual solution..
<br\>
b. using pivoting

In [102]:
import numpy as np

j=[0.143,0.357,2.01,-5.173]
k=[-1.31,.911,1.99,-5.458]
l=[11.2,-4.3,-.605,4.415]

print('initial matrix-  \n ' + str(np.asmatrix([j,k,l])) + 
      '\n we need to swap the pivot row (R3) with R1  \n ')

l=[0.143,0.357,2.01,-5.173]
k=[-1.31,.911,1.99,-5.458]
j=[11.2,-4.3,-.605,4.415]

print(" ... \n " + str(np.asmatrix([j,k,l])))

j=[round((1/11.2)*i,3) for i in j]
k=[round((1/-1.31)*i,3) for i in k]
l=[round((1/.143)*i,3) for i in l]

print(" \n getting 1's in left most column -  \n " + str(np.asmatrix([j,k,l])))
k=np.subtract(k,j)
l=np.subtract(l,j)
print("subtracting first row form second and third-   \n " + str(np.asmatrix([j,k,l])) +
      "\n before continuing with 2nd column, need to swap R2 and R3 \n ")

a=k
b=l
l=a
k=b

print("...   \n " + str(np.asmatrix([j,k,l])))

k=[round((1/2.881)*i,3) for i in k]
l=[round((1/-0.311)*i,3) for i in l]

print("continuing with 2nd column  \n " + str(np.asmatrix([j,k,l])))
l=np.subtract(l,k)
l=[round((1/-0.187)*i,3) for i in l]
print( "... \n" + str(np.asmatrix([j,k,l])) + " \n now we have all 1's on diagonal so we will start back substitution \n ")

l_alpha=[round((-4.898)*i,3) for i in l]
k=np.add(k,l_alpha)
l_beta=[round((.054)*i,3) for i in l]
j=np.add(j,l_beta)
print("R2 + -4.898*R3   \n  R1 + .054*R3 \n" + str(np.asmatrix([j,k,l])))
k_alpha=[round((0.384)*i,3) for i in k]
j=np.add(j,k_alpha)
print("R1 + 0.384*R2   \n " + str(np.asmatrix([j,k,l])))

initial matrix-  
 [[  0.143   0.357   2.01   -5.173]
 [ -1.31    0.911   1.99   -5.458]
 [ 11.2    -4.3    -0.605   4.415]]
 we need to swap the pivot row (R3) with R1  
 
 ... 
 [[ 11.2    -4.3    -0.605   4.415]
 [ -1.31    0.911   1.99   -5.458]
 [  0.143   0.357   2.01   -5.173]]
 
 getting 1's in left most column -  
 [[  1.     -0.384  -0.054   0.394]
 [  1.     -0.695  -1.519   4.166]
 [  1.      2.497  14.056 -36.175]]
subtracting first row form second and third-   
 [[  1.     -0.384  -0.054   0.394]
 [  0.     -0.311  -1.465   3.772]
 [  0.      2.881  14.11  -36.569]]
 before continuing with 2nd column, need to swap R2 and R3 
 
...   
 [[  1.     -0.384  -0.054   0.394]
 [  0.      2.881  14.11  -36.569]
 [  0.     -0.311  -1.465   3.772]]
continuing with 2nd column  
 [[  1.     -0.384  -0.054   0.394]
 [  0.      1.      4.898 -12.693]
 [ -0.      1.      4.711 -12.129]]
... 
[[  1.     -0.384  -0.054   0.394]
 [  0.      1.      4.898 -12.693]
 [  0.     -0.      1.    

Here we have a much better solution [1.029,2.079,-3.016], notice the multipliers in the back substitution were much smaller (the goal of pivoting) which causes less change in the augmented vector which in turn causes less error due to rounding.

## 2. Ill-Conditioned Systems

Consider the linear system with

$$
Ax = b
$$

with

$$
A =
\begin{bmatrix}
1& 1 \\
1& 1.001
\end{bmatrix}
$$

and $b$ arbitrary. Note that $A$ is very close to being non-invertible, a sure sign of ill-conditioned systems (small perturbations in $b$ can lead to large perturbations in the solution).

### (a) Calculate the condition number of $A$ using the max norm for matrices.

Max norm is just a the max row sum of $A$.

The condition number is less than or equal to $||A||*||A^{-1}||$

Which is:

$2.001 * 2001 = 4002.001$

### (b)

Intuition:

The solution to y represents the case where we perturb b by a small amount. Because this system is relatively ill-conditioned, then the difference between x and y should be much larger than the magnitude of the perturbance.

Demostration:

Let $b = \begin{bmatrix}1 \\ 0 \end{bmatrix}$ and $\epsilon = \begin{bmatrix}0.01 \\ -0.01 \end{bmatrix}$


Then $x = A^{-1} * b = \begin{bmatrix}1001 \\ -1000 \end{bmatrix}$

and $y = A^{-1} * (b + \epsilon) = \begin{bmatrix}1021.01 \\ -1020 \end{bmatrix}$

now $||y-x||_{\infty} = 20.01$ and $||\epsilon||_{\infty} = .01$. Clearly this is a large difference, and supports the behavior expected by my intuition.
